src/Targets/Microsoft.Managed.DesignTime.targets

```diff
@@ -283,4 +278,15 @@
   <!-- This target collects all the resolved references that are used to actually compile. -->
   <Target Name="CollectResolvedCompilationReferencesDesignTime" DependsOnTargets="CompileDesignTime" Returns="@(ReferencePathWithRefAssemblies)" />
 
+  <!-- This target collects all up-to-date file additions. -->
+  <Target Name="CollectUpToDateInputsDesignTime" DependsOnTargets="CompileDesignTime" Returns="@(UpToDateCheckInput)">
+    <!-- If the project we're compiling has a ruleset, make sure we check it for up-to-date checks. -->
+    <ItemGroup Condition="'$(ResolvedCodeAnalysisRuleSet)' != ''">
+      <UpToDateCheckInput Include="$([System.IO.Path]::GetFullPath($(ResolvedCodeAnalysisRuleSet)))" />
```

Comment:
Design-Time builds are not free and add a non-trivial overhead. The first time you hook up the event, that is going to kick of a build just for you because we haven't yet got the data. I plan on making a change where we make sure by the very first design-time build we kick off, we've gather the first set of data of all the services in the system so that don't have them trickling in, kicking off extra underneeded builds. That isn't in, yet.

I've look at your code, I I see that you were already waiting on a design-time build (JointRuleSource) before your data was populated, it looks like a race:

- I make a change
- I build immediately before the design-time build has occurred
- The data you are looking at is out of date

Or is the design-time build a critical task that you bail on?